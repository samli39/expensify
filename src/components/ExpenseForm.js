import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from'react-dates';



//parent:AddExpensePage,EditExpensePage

class ExpenseForm extends React.Component{
	state={
		description:this.props.expense?this.props.expense.description:"",
		amount:this.props.expense ? (this.props.expense.amount/100).toString():"",
		note:this.props.expense ? this.props.expense.note : "",
		createdAt:this.props.expense?moment(this.props.expense.createdAt): moment(),
		calenderFocused:false,
		error:""
	}
	//form submition
	onSubmit=(e)=>{
		e.preventDefault();
		if(!this.state.description || !this.state.amount){
			this.setState(()=>({error:"please enter description or amount"}))
		}else{
			this.setState(()=>({error:""}));
			this.props.onSubmit({
				description:this.state.description,
				amount:parseFloat(this.state.amount,10)*100,
				createdAt:this.state.createdAt.valueOf(),
				note:this.state.note
			})
		}
	}
	//description
	onDescriptionChange=(e)=>{
		const description=e.target.value;
		this.setState(()=>({description}));
	}

	//Amount
	onAmountChange=(e)=>{
		const amount=e.target.value;
		//!amount refer to delete amount
		//^\d{1,}-start from number
		//{1,}-at least 1 to infinite
		//()?-appear zero or one time
		// \. - .symbol
		//\d{0,2} - happen 0 to 2 times
		( !amount || (/^\d{1,}(\.\d{0,2})?$/).test(amount))
		&&
		this.setState(()=>({amount}));
	}
	
	//date picker
	onDateChange=(createdAt)=>{
		if(createdAt)
			this.setState(()=>({createdAt}));
	}
	
	//focus on date picker
	onFocusChange=({focused})=>{
		this.setState(()=>({calenderFocused:focused}));
	}
	//note
	onNoteChange=(e)=>{
		const note = e.target.value;
		this.setState(()=>({note}));
	}
	
	render(){
		return(
				<form className="form" onSubmit={this.onSubmit}>
					{this.state.error ? 
						<p className="form__error">{this.state.error}</p>
					:
						<p> Please provide description and amount </p>
					}

					{/*description*/}
					<input 
						className="text-input"
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>

					{/*amount*/}
					<input 
						type="text"
						className="text-input"
						placeholder="Amount"
						onChange={this.onAmountChange}
						value={this.state.amount}
					/>
				{/*	dates
					numberOfMonths =show how many month
					isOutsideRange =allow pick previous month
				*/}
				<SingleDatePicker 
					date={this.state.createdAt}
					onDateChange={this.onDateChange}
					focused={this.state.calenderFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={()=>false}
					showClearDate={true}
				/>


					
					{/* note */}
					<textarea
						className="textarea"
						placeholder="Add some not for your expense{optional}"
						value = {this.state.note}
						onChange={this.onNoteChange}
					>

					</textarea>
					<div className="form__button">
						<button className="button">Add Expense</button>
					</div>
				</form>
		)
	}
}

export default ExpenseForm;