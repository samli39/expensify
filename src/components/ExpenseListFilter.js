import React from 'react';
import {connect} from 'react-redux';
import {setFilter,sortByDate,sortByAmount,setStartDate,setEndDate} from '../action/action_filter';
import {DateRangePicker} from 'react-dates';

//parent:ExpenseDashboardPahe

export class ExpenseListFilter extends React.Component{
	state={
		calendarFocused:null
	}
	
	// input text for filter text
	onTextChange=(e)=>{
		this.props.setFilter(e.target.value);
	}

	//for sortBy
	onSortByChange=(e)=>{
		(e.target.value === "date") ?
		this.props.sortByDate()
		:
		this.props.sortByAmount();
	}

	//DateRanger Picker
	onDateChange=({startDate,endDate})=>{
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}

	onFocusChange=(calendarFocused)=>{
		this.setState(()=>({calendarFocused}))
	}

	render(){
		return (
			<div className="content-container">
				<div className="input-group">
					<div className="input-group__item">
						{/*for filter text*/}
						<input 
							type="text" 
							className="text-input"
							placeholder="Search expenses"
							value={this.props.filters.text}
							onChange={this.onTextChange}
						/>
					</div>
					<div className="input-group__item">
						{/*option for sortBy*/}
						<select
							className="select"
							value={this.props.filters.sortBy}
							onChange={this.onSortByChange}
							>
							<option value="date">Date</option>
							<option value="amount">Amount</option>
						</select>
					</div>
					<div className="input-group__item">
						<DateRangePicker
							startDate={this.props.filters.startDate}
							endDate={this.props.filters.endDate}
							onDatesChange={this.onDateChange}
							focusedInput={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={()=>false}
							showClearDates={true}
						/>
					</div>
				</div>
			</div>

		)
	}
}

const mapStateToProps=(state)=>({
	filters:state.filters
})
const mapDispatchToProps=(dispatch)=>({
	setFilter:(text)=>dispatch(setFilter({text})),
	sortByDate:()=>dispatch(sortByDate()),
	sortByAmount:()=>dispatch(sortByAmount()),
	setStartDate:(date)=>dispatch(setStartDate(date)),
	setEndDate:(date)=>dispatch(setEndDate(date))
})

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter);
// export default ExpenseListFilter;

