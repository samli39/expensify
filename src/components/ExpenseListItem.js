import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
//parent:ExpenseList

const ExpenseListItem=({id,description,createdAt,amount})=>(
	<div>
		<Link className="list-item" to ={`/edit/${id}`}>
			<div className="list-item__header">
				<h3 className="list-item__title">{description}</h3>
				<span className="list-item__sub-title">{ moment(createdAt).format("MMMM Do,YYYY") }</span>
			</div>
			<div>
				<h3 className="list-item__data">{numeral(amount/100).format('$0,0.00')}</h3>
			</div>
		</Link>
	</div>
)



export default ExpenseListItem;