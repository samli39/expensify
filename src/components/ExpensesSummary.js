import React from 'react';
import {connect} from 'react-redux';
import expenseSelector from '../selector/expenseSelector';
import numeral from 'numeral';
import expenseTotal from '../selector/expenseTotal';

//parent:ExpenseDashboardPage
export const ExpensesSummary=(props)=>{
	const amount = expenseTotal(props.expenses);
	const expenseWord = (props.expenses.length === 1) ? 'expense' :'expenses';
	return(
		<div>
			<h1>Viewing {props.expenses.length} {expenseWord} totalling {numeral(amount/100).format('$0,0.00')}</h1>
		</div>
	)
}

const mapStateToProps=(state)=>({
	expenses:expenseSelector(state.expenses,state.filters)
});

//other way

// export const ExpensesSummary =({expenseCount,expenseTotal})=>(
// 	<div>
// 	<h1>Viewing {expenseCount} expenses totalling : {expenseTotal}</h1>
// 	</div>
// 	)




// const mapStateToProps=(state)=>{
// 	const expenses=expenseSelector(state.expenses,state.filters);
// 	return{
// 		expenseCount:expenses.length,
// 		expenseTotal:numeral(getExpensesTotal(expenses)).format("$0,0.00")
// 	}
// }

export default connect(mapStateToProps)(ExpensesSummary);