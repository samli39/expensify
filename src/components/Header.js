import React from 'react';
import {connect} from 'react-redux';
import {NavLink,withRouter} from 'react-router-dom';
import {startLogout} from '../action/action_auth';

export const Header =(props)=>{
	const logout=()=>{
		props.startLogout()
	}
	return(
		<header>
			<h1>Expensify</h1>
			<NavLink to="/dashboard" activeClassName="is-actives" >Home </NavLink>
			<br />
			<NavLink to="/create" activeClassName="is-actives">create </NavLink>
			<br />
			<button onClick={logout}>Logout</button>
		</header>
	)
}
const mapDispatchToProps=(dispatch)=>{
	return{
		startLogout:()=>dispatch(startLogout())
	}
}
export default connect(undefined,mapDispatchToProps)(Header);