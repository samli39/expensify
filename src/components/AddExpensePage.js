import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startAddExpense} from '../action/action_expenses'


export const AddExpensePage=(props)=>{
	const onSubmit=(expense)=>{
		props.startAddExpense(expense);
		props.history.push("/dashboard");
	}


	return(
		<div>
			it is add expense Page
			<ExpenseForm onSubmit={onSubmit}/>
		</div>

	)
}

const mapDispatchToProps=(dispatch)=>({
	startAddExpense:(expense)=>dispatch(startAddExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage);