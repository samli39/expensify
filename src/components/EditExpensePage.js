import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpense,startRemoveExpense} from '../action/action_expenses';
import {Redirect} from 'react-router-dom';

export const EditExpensePage=(props)=>{
	const onSubmit=(expense)=>{
		props.startEditExpense(props.expense.id,expense).then(()=>{
			props.history.push("/dashboard");
		});
	}

	const remove=()=>{
		props.startRemoveExpense(props.expense.id).then(()=>{
			props.history.push('/dashboard');
		});
	}

	return(
		!props.expense ?(
			<Redirect to="/notFountdPage" />
		):(
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="paage-header__title">Edit Expense</h1> 
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm onSubmit={onSubmit} expense={props.expense}/>
					<button className="button button--remove" onClick={remove} > Remove Expense </button>
				</div>
			</div>
		)
	)
}

const mapStateToProps=(state,props)=>({
	expense:state.expenses.find((expense)=>{
		return expense.id === props.match.params.id;
	})
})

const mapDispatchToProps=(dispatch)=>({
	startEditExpense:(id,update)=>dispatch(startEditExpense(id,update)),
	startRemoveExpense:(id)=>dispatch(startRemoveExpense(id))
})

//or

// const mapDispatchToProps=(dispatch,props)=>({
// 	editExpense:(id,update)=>dispatch(editExpense(id,update)),
// 	removeExpense:()=>dispatch(removeExpense(props.expense.id))
// })
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);