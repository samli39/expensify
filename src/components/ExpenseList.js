import React from 'react';
import {connect} from 'react-redux';
import expenseSelector from '../selector/expenseSelector';
import ExpenseListItem from './ExpenseListItem';

//parent:ExpenseDashboardPage
export const ExpenseList = (props)=>(
	<div>
		<h1>Expense List </h1>
		{
			( props.expenses.length > 0 )?(
				props.expenses.map((e)=>
						<ExpenseListItem key={e.id} {...e} />
					)
				):(
					<p> there is no item </p>
				)

		}
	</div>
)

const mapStateToProps=(state)=>({
	expenses:expenseSelector(state.expenses,state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
