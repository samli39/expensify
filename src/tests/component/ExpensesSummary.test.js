import React from 'react';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixture/expenses';
import {shallow} from 'enzyme';

test('should render ExpensesSummary correctly with no expenses',()=>{

	const wrapper =shallow(<ExpensesSummary expenses={[]}/>);
	expect(wrapper).toMatchSnapshot();
})

test('should show 1 expense and total $1.95',()=>{

	const wrapper =shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
	expect(wrapper).toMatchSnapshot();
})

test('should show three expenses and total $1,141.95',()=>{

	const wrapper =shallow(<ExpensesSummary expenses={expenses}/>);
	expect(wrapper).toMatchSnapshot();
})