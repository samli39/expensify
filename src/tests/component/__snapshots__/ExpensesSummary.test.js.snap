// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with no expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h1>
</div>
`;

exports[`should show 1 expense and total $1.95 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
  </h1>
</div>
`;

exports[`should show three expenses and total $1,141.95 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
  </h1>
</div>
`;
